### ISSUE TITLE
Changing direct routing device ifindex causes datapath compilation failures

### ISSUE THREAD

**Issue by jerry2714** [cilium/cilium#29568]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I'm using Netmaker (Wireguard based VPN) as the underlying network of my k8s cluster, I found  sometimes  any new pod stuck in `ContainerCreating` state until I restart cilium pod on the  node. The log shows bpf template complilation failed due to 'DIRECT_ROUTING_DEV_IFINDEX' macro redefined.

I tried to repair this situation by editing source file in the cilium pod, for example, the log shows like this:
```log
2023-12-03T08:11:43.281638736Z level=warning msg="/var/run/cilium/state/globals/node_config.h:44:9: error: 'DIRECT_ROUTING_DEV_IFINDEX' macro redefined [-Werror,-Wmacro-redefined]" subsys=datapath-loader
2023-12-03T08:11:43.281688517Z level=warning msg="#define DIRECT_ROUTING_DEV_IFINDEX 38" subsys=datapath-loader
2023-12-03T08:11:43.282142742Z level=warning msg="        ^" subsys=datapath-loader
2023-12-03T08:11:43.282158954Z level=warning msg="/var/run/cilium/state/templates/34b25ed53ec8574840c8f546fddf6dccf84a43fadbeb7ea8bca247fac487eb67/ep_config.h:20:9: note: previous definition is here" subsys=datapath-loader
2023-12-03T08:11:43.282168361Z level=warning msg="#define DIRECT_ROUTING_DEV_IFINDEX 54" subsys=datapath-loader
```

and I execute `ip a` on my worker node, it shows the netmaker interface index is 54
```sh
54: netmaker: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1420 qdisc noqueue state UNKNOWN group default qlen 1000
    link/none 
    inet 10.101.0.4/16 brd 10.101.255.255 scope global netmaker
       valid_lft forever preferred_lft forever
```
so I use `sed` to replace the index in `/var/run/cilium/state/globals/node_config.h` in that cilium pod

```sh
sed -i  's/DIRECT_ROUTING_DEV_IFINDEX 38/DIRECT_ROUTING_DEV_IFINDEX 54/g' /var/run/cilium/state/globals/node_config.h
```
And after a few seconds, the stuck pod successfully created, so I think the problem is that netmaker somehow  
recreate the network interface, and cilium didn't detect and change the code while network interface changed.

I'm not sure is it a bug or just because some misconfiguration, and currently I have no idea what is the best way to  workaround, here is my helm chart values, please give me some advice, thank you!

```yaml
rollOutCiliumPods: true
k8sServiceHost: 10.101.0.3
k8sServicePort: 6443
kubeProxyReplacement: strict
kubeProxyReplacementHealthzBindAddr: "0.0.0.0:10256"
ipam:
  operator:
    clusterPoolIPv4PodCIDRList:
      - 10.244.0.0/16
bpf:
  masquerade: true
hubble:
  ui:
    enabled: true
    standalone:
      enabled: true 
```

### Cilium Version

1.14.4

### Kernel Version

5.15.0-89-generic

### Kubernetes Version

1.28.4

### Relevant log output

```shell
2023-12-03T08:11:43.281068642Z level=error msg="Failed to compile bpf_lxc.o: exit status 1" compiler-pid=1634 linker-pid=1635 subsys=datapath-loader
2023-12-03T08:11:43.281129744Z level=warning msg="In file included from /var/lib/cilium/bpf/bpf_lxc.c:9:" subsys=datapath-loader
2023-12-03T08:11:43.281638736Z level=warning msg="/var/run/cilium/state/globals/node_config.h:44:9: error: 'DIRECT_ROUTING_DEV_IFINDEX' macro redefined [-Werror,-Wmacro-redefined]" subsys=datapath-loader
2023-12-03T08:11:43.281688517Z level=warning msg="#define DIRECT_ROUTING_DEV_IFINDEX 38" subsys=datapath-loader
2023-12-03T08:11:43.282142742Z level=warning msg="        ^" subsys=datapath-loader
2023-12-03T08:11:43.282158954Z level=warning msg="/var/run/cilium/state/templates/34b25ed53ec8574840c8f546fddf6dccf84a43fadbeb7ea8bca247fac487eb67/ep_config.h:20:9: note: previous definition is here" subsys=datapath-loader
2023-12-03T08:11:43.282168361Z level=warning msg="#define DIRECT_ROUTING_DEV_IFINDEX 54" subsys=datapath-loader
2023-12-03T08:11:43.282177356Z level=warning msg="        ^" subsys=datapath-loader
2023-12-03T08:11:43.282186062Z level=warning msg="1 error generated." subsys=datapath-loader
2023-12-03T08:11:43.282194973Z level=warning msg="JoinEP: Failed to compile" debug=true error="Failed to compile bpf_lxc.o: exit status 1" params="&{Source:bpf_lxc.c Output:bpf_lxc.o OutputType:obj Options:[]}" subsys=datapath-loader
2023-12-03T08:11:43.282225804Z level=error msg="BPF template object creation failed" bpfHeaderfileHash=34b25ed53ec8574840c8f546fddf6dccf84a43fadbeb7ea8bca247fac487eb67 error="failed to compile template program /var/run/cilium/state/templates/34b25ed53ec8574840c8f546fddf6dccf84a43fadbeb7ea8bca247fac487eb67: Failed to compile bpf_lxc.o: exit status 1" subsys=datapath-loader
2023-12-03T08:11:43.282889765Z level=error msg="Error while rewriting endpoint BPF program" containerID=516aa25493 datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=2702 error="BPF template compilation failed: failed to compile template program /var/run/cilium/state/templates/34b25ed53ec8574840c8f546fddf6dccf84a43fadbeb7ea8bca247fac487eb67: Failed to compile bpf_lxc.o: exit status 1" identity=32388 ipv4=10.244.2.25 ipv6= k8sPodName=default/multitool-57888df9c8-k7g9x subsys=endpoint
2023-12-03T08:11:43.282928766Z level=info msg="generating BPF for endpoint failed, keeping stale directory" containerID=516aa25493 datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=2702 error="BPF template compilation failed: failed to compile template program /var/run/cilium/state/templates/34b25ed53ec8574840c8f546fddf6dccf84a43fadbeb7ea8bca247fac487eb67: Failed to compile bpf_lxc.o: exit status 1" file-path=2702_next_fail identity=32388 ipv4=10.244.2.25 ipv6= k8sPodName=default/multitool-57888df9c8-k7g9x subsys=endpoint
2023-12-03T08:11:43.283422321Z level=warning msg="Regeneration of endpoint failed" bpfCompilation=231.406836ms bpfLoadProg=0s bpfWaitForELF=231.885304ms bpfWriteELF=0s containerID=516aa25493 datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=2702 error="BPF template compilation failed: failed to compile template program /var/run/cilium/state/templates/34b25ed53ec8574840c8f546fddf6dccf84a43fadbeb7ea8bca247fac487eb67: Failed to compile bpf_lxc.o: exit status 1" identity=32388 ipv4=10.244.2.25 ipv6= k8sPodName=default/multitool-57888df9c8-k7g9x mapSync="79.253µs" policyCalculation="65.473µs" prepareBuild="765.22µs" proxyConfiguration="11.947µs" proxyPolicyCalculation="23.268µs" proxyWaitForAck=0s reason="updated security labels" subsys=endpoint total=409.3182ms waitingForCTClean=175.505466ms waitingForLock="1.213µs"
2023-12-03T08:11:43.283456810Z level=error msg="endpoint regeneration failed" containerID=516aa25493 datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=2702 error="BPF template compilation failed: failed to compile template program /var/run/cilium/state/templates/34b25ed53ec8574840c8f546fddf6dccf84a43fadbeb7ea8bca247fac487eb67: Failed to compile bpf_lxc.o: exit status 1" identity=32388 ipv4=10.244.2.25 ipv6= k8sPodName=default/multitool-57888df9c8-k7g9x subsys=endpoint
```

### Anything else?

The relevant event of the pod that cannot be created:
```
Warning  FailedCreatePodSandBox  16m    kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "46f69a0c3294b310c825f5ad2e15e367608f6bf732d5a496aea4e9e32a9f93dd": plugi │
│ n type="cilium-cni" failed (add): unable to create endpoint: Cilium API client timeout exceeded
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ti-mo** [comment]
@jerry2714 Thanks for the detailed report! I don't think Cilium currently accounts for the direct routing ifindex changing in general. If this interface goes away after starting the agent, a restart will be necessary to add the necessary routes and other bpf-related objects to the device. (the host datapath needs to be regenerated)

We've been working on making this part of the stack more reactive to changes, but as usual this requires significant changes to the agent's internals. I'll keep this around as a data point to feed into this work.

**Comment by apolovov** [comment]
The issue is relevant for a case:
* cluster network interface is vlan1000, which is configured on top of bond2 interface,
* the interface is configured via `/etc/network/interfaces`,
* after changing the configuration, ops apply it with `systemctl restart networking.service` and interfaces are being recreated,
* so, the ifindexes become new.

before:
```
1074: vlan1000@bond2
```
after:
```
1076: vlan1000@bond2
```

**Comment by julianwiedmann** [comment]
Cilium v1.19 brings https://github.com/cilium/cilium/pull/40430, which should avoid these compilation problems.

Still no dynamic reconciliation of the ifindex though.

**Comment by joestringer** [comment]
@julianwiedmann what is your intent around assigning this issue to the 1.19 milestone? Note that only one milestone can be assigned, so if this is not progressing to be resolved on the 1.19 timeline then we will likely just end up removing the 1.19 milestone from the issue.

**Comment by julianwiedmann** [comment]
> [@julianwiedmann](https://github.com/julianwiedmann) what is your intent around assigning this issue to the 1.19 milestone? Note that only one milestone can be assigned, so if this is not progressing to be resolved on the 1.19 timeline then we will likely just end up removing the 1.19 milestone from the issue.

See the comment right above yours.

**Comment by joestringer** [comment]
Let me rephrase: By setting the 1.19 milestone, are you asserting that you will work on this to solve it for v1.19 final or are you aware of other developers who will do so?

**Comment by ti-mo** [comment]
@joestringer My interpretation: I think the rationale is that this particular case will be fixed in 1.19, and a backport to 1.18 or earlier is not on the table since the changes are not backwards-compatible.

@julianwiedmann Following that reasoning, should we close this as fixed?

**Comment by joestringer** [comment]
I see. The linked commit was merged months ago so it should already be part of the v1.19 prerelease builds. Given that, I would close the issue with a note that the problem will be solved in v1.19 (ETA next month).

> Still no dynamic reconciliation of the ifindex though.

This was the comment that confused me - it sounds like there is remaining work to solve the bug. But if this is extra tech debt work that is not directly related to solving the bug reported here, maybe we can track that separately.

**Comment by julianwiedmann** [comment]
Closing as per ☝️.
